db/system_keyspace.cc

```diff
@@ -2887,23 +2887,6 @@ static std::set<sstring> decode_features(const set_type_impl::native_type& featu
     return fset;
 }
 
-static bool must_have_tokens(service::node_state nst) {
```

Comment:
Why do we delete it instead of checking that the node is non zero token one?

Reply:
Zero-token nodes can be in almost any state (the only exception is `rebuilding` because we skip the rebuild operation for it). For example, if a node is in the state `removing` and it doesn't have tokens, we must assume that it's correct and this node is a zero-token node. So, for any state other that `rebuilding`, `must_have_tokens` must return `false`, which makes this function almost redundant.

Reply:
That's because we lose the information that the node is zero-token node after join. We rely on tokens to be empty to detect it, but this makes all the rest of the code less safe because now we cannot check some invariants any more

Reply:
> we cannot check some invariants any more

They just stopped being invariants.

Previously it was true that `node is in normal state => it has tokens`, now it is no longer true -- you could artificially introduce invariants by introducing a bool like `is_zero_token` and then we would have `not is_zero_token and node is in normal state => it has tokens` but that would just make the code more complex.

Consider what would happen if we implemented the system from scratch, supporting both zero-token and token-owning nodes from the start. Would we introduce a bool like that?

Reply:
> > we cannot check some invariants any more
> 
> They just stopped being invariants.

Yes. And by doing that they making it harder to detect bugs.

> 
> Previously it was true that `node is in normal state => it has tokens`, now it is no longer true -- you could artificially introduce invariants by introducing a bool like `is_zero_token` and then we would have `not is_zero_token and node is in normal state => it has tokens` but that would just make the code more complex.
> 
> Consider what would happen if we implemented the system from scratch, supporting both zero-token and token-owning nodes from the start. Would we introduce a bool like that?

We would make it possible to distinguish between empty set and non existing set in the topology state and would have made ring optional all the same. We are discussing a bool only because of the weird set properties.

Reply:
@gleb-cloudius I did what you asked for. It turned out that it's much easier than we thought. We don't even need a new column.

The join procedure ensures that a joining node is zero-token <=> it sends a request to the join coordinator with `num_tokens=0` and `tokens_string=""`. This allows the join coordinator to know if a joining node is zero-token, and it can, for example, check that a replacing zero-token node replaces another zero-token node. This logic was in old version of the PR.

What I forgot about i that the join coordinator adds `num_tokens` and `tokens_string` to `system.topology` and these values never change. Therefore, `system_keyspace::load_topology_state` can distinguish zero-token nodes and token-owning nodes in the same way -- it just needs to look at `num_tokens` and `tokens_string`.

So, I reverted the patch removing the optional and only added a few lines to `system_keyspace::load_topology_state`.